<a href="https://colab.research.google.com/github/matthewreed2000/cse380-notebooks/blob/master/09_4_Expound_on_Topics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Expound on Topics
## Wednesday, 3 March 2021

Let's explore prime sieving a bit, shall we?

In [1]:
# From the
# Exemplary answer to Exercise 307. The magic happens
# in realprimes, which is based on a sieve found
# somewhere online. This takes a few  seconds to run.
# Pretty close to as fast as (or is it faster than?)
# using the highly tuned primerange or sieve functions
# found in the sympy module.
def realprimes(n):
  """Input n > 5, returns a list of primes, 3 < p < n.
  """
  if n <= 5:
      raise ValueError(f"realprimes needs a number > 5, but got {n}.")
  n, correction = n - n % 6 + 6, 2 - (n % 6 > 1)
  sieve = [True] * (n // 3)
  for i in range(1, int(n ** 0.5) // 3 + 1):
      if sieve[i]:
          k = 3 * i + 1 | 1
          sieve[k * k // 3 :: 2 * k] = [False] * ((n // 6 - k * k // 6 - 1) // k + 1)
          sieve[k * (k - 2 * (i & 1) + 4) // 3 :: 2 * k] = [False] * ((n // 6 - k * (k - 2 * (i & 1) + 4) // 6 - 1) // k + 1)
  return [3 * i + 1 | 1 for i in range(1, n // 3 - correction) if sieve[i]]

## How about this?

In [2]:
def repres(rows, *moduli):
  """Repeating residues made
     by modding range(rows+1)
     by each of an arbitrary
     number of moduli, one
     tuple of residues per row.
  """
  for n in range(rows + 1):
    print(n, end=' ')
    residues=list(map(lambda m:\
      n % m, moduli))
    print(*residues)

import functools, math, operator, sys

if sys.version_info >= (3, 8):
  product = math.prod

  def lcm(*numbers):
    return math.prod(numbers) // math.gcd(*numbers)
else:
  product = functools.partial(functools.reduce, operator.mul)
  greatest_common_divisor = functools.partial(functools.reduce, math.gcd)
  def lcm(*numbers):
    return product(numbers) // greatest_common_divisor(numbers)

def represBG(*moduli):
  """Repeating residues with
     bijectivity guaranteed
     made by modding
     range(lcm(*moduli))
     by each of an arbitrary
     number of moduli, one
     tuple of residues per row.
  """
  for n in range(lcm(*moduli)):
    # print(n, end=' <--> (')
    residues=list(map(lambda m:\
      n % m, moduli))
    # print(*residues, sep=', ', \
    #                  end=')\n')
    yield residues

## TODO for ICE
Write a good filter to tell primes from nonprimes given their residue lists:

In [3]:
2 * 3 * 5 * 7 * 11

2310

In [4]:
residues = [(x,y) for x, y in zip(range(2310),represBG(2, 3, 5, 7, 11))]
residues

[(0, [0, 0, 0, 0, 0]),
 (1, [1, 1, 1, 1, 1]),
 (2, [0, 2, 2, 2, 2]),
 (3, [1, 0, 3, 3, 3]),
 (4, [0, 1, 4, 4, 4]),
 (5, [1, 2, 0, 5, 5]),
 (6, [0, 0, 1, 6, 6]),
 (7, [1, 1, 2, 0, 7]),
 (8, [0, 2, 3, 1, 8]),
 (9, [1, 0, 4, 2, 9]),
 (10, [0, 1, 0, 3, 10]),
 (11, [1, 2, 1, 4, 0]),
 (12, [0, 0, 2, 5, 1]),
 (13, [1, 1, 3, 6, 2]),
 (14, [0, 2, 4, 0, 3]),
 (15, [1, 0, 0, 1, 4]),
 (16, [0, 1, 1, 2, 5]),
 (17, [1, 2, 2, 3, 6]),
 (18, [0, 0, 3, 4, 7]),
 (19, [1, 1, 4, 5, 8]),
 (20, [0, 2, 0, 6, 9]),
 (21, [1, 0, 1, 0, 10]),
 (22, [0, 1, 2, 1, 0]),
 (23, [1, 2, 3, 2, 1]),
 (24, [0, 0, 4, 3, 2]),
 (25, [1, 1, 0, 4, 3]),
 (26, [0, 2, 1, 5, 4]),
 (27, [1, 0, 2, 6, 5]),
 (28, [0, 1, 3, 0, 6]),
 (29, [1, 2, 4, 1, 7]),
 (30, [0, 0, 0, 2, 8]),
 (31, [1, 1, 1, 3, 9]),
 (32, [0, 2, 2, 4, 10]),
 (33, [1, 0, 3, 5, 0]),
 (34, [0, 1, 4, 6, 1]),
 (35, [1, 2, 0, 0, 2]),
 (36, [0, 0, 1, 1, 3]),
 (37, [1, 1, 2, 2, 4]),
 (38, [0, 2, 3, 3, 5]),
 (39, [1, 0, 4, 4, 6]),
 (40, [0, 1, 0, 5, 7]),
 (41, [1, 2, 1, 6, 8]),

In [5]:
[x for x in residues if x[1].count(0) == 0]

[(1, [1, 1, 1, 1, 1]),
 (13, [1, 1, 3, 6, 2]),
 (17, [1, 2, 2, 3, 6]),
 (19, [1, 1, 4, 5, 8]),
 (23, [1, 2, 3, 2, 1]),
 (29, [1, 2, 4, 1, 7]),
 (31, [1, 1, 1, 3, 9]),
 (37, [1, 1, 2, 2, 4]),
 (41, [1, 2, 1, 6, 8]),
 (43, [1, 1, 3, 1, 10]),
 (47, [1, 2, 2, 5, 3]),
 (53, [1, 2, 3, 4, 9]),
 (59, [1, 2, 4, 3, 4]),
 (61, [1, 1, 1, 5, 6]),
 (67, [1, 1, 2, 4, 1]),
 (71, [1, 2, 1, 1, 5]),
 (73, [1, 1, 3, 3, 7]),
 (79, [1, 1, 4, 2, 2]),
 (83, [1, 2, 3, 6, 6]),
 (89, [1, 2, 4, 5, 1]),
 (97, [1, 1, 2, 6, 9]),
 (101, [1, 2, 1, 3, 2]),
 (103, [1, 1, 3, 5, 4]),
 (107, [1, 2, 2, 2, 8]),
 (109, [1, 1, 4, 4, 10]),
 (113, [1, 2, 3, 1, 3]),
 (127, [1, 1, 2, 1, 6]),
 (131, [1, 2, 1, 5, 10]),
 (137, [1, 2, 2, 4, 5]),
 (139, [1, 1, 4, 6, 7]),
 (149, [1, 2, 4, 2, 6]),
 (151, [1, 1, 1, 4, 8]),
 (157, [1, 1, 2, 3, 3]),
 (163, [1, 1, 3, 2, 9]),
 (167, [1, 2, 2, 6, 2]),
 (169, [1, 1, 4, 1, 4]),
 (173, [1, 2, 3, 5, 8]),
 (179, [1, 2, 4, 4, 3]),
 (181, [1, 1, 1, 6, 5]),
 (191, [1, 2, 1, 2, 4]),
 (193, [1, 1, 3, 4,

This fails on numbers whose factors are primes outside of (2,3,5,7,11), such as 169 (which is 13*13).

In [2]:
def primes_to_max(upto):
  moduli = []
  for n in range(2,upto+1):
    residues=list(map(lambda m:\
      n % m, moduli))
    if residues.count(0) != 0:
      continue
    moduli.append(n)
    yield n

In [3]:
[*primes_to_max(2310)]

[2,
 3,
 5,
 7,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97,
 101,
 103,
 107,
 109,
 113,
 127,
 131,
 137,
 139,
 149,
 151,
 157,
 163,
 167,
 173,
 179,
 181,
 191,
 193,
 197,
 199,
 211,
 223,
 227,
 229,
 233,
 239,
 241,
 251,
 257,
 263,
 269,
 271,
 277,
 281,
 283,
 293,
 307,
 311,
 313,
 317,
 331,
 337,
 347,
 349,
 353,
 359,
 367,
 373,
 379,
 383,
 389,
 397,
 401,
 409,
 419,
 421,
 431,
 433,
 439,
 443,
 449,
 457,
 461,
 463,
 467,
 479,
 487,
 491,
 499,
 503,
 509,
 521,
 523,
 541,
 547,
 557,
 563,
 569,
 571,
 577,
 587,
 593,
 599,
 601,
 607,
 613,
 617,
 619,
 631,
 641,
 643,
 647,
 653,
 659,
 661,
 673,
 677,
 683,
 691,
 701,
 709,
 719,
 727,
 733,
 739,
 743,
 751,
 757,
 761,
 769,
 773,
 787,
 797,
 809,
 811,
 821,
 823,
 827,
 829,
 839,
 853,
 857,
 859,
 863,
 877,
 881,
 883,
 887,
 907,
 911,
 919,
 929,
 937,
 941,
 947,
 953,
 967,
 971,
 977,
 983,
 991,
 997,
 1009,
 1013,
 1019,


# TODO for ICE

Write a Python function to restore the list of the first 10 million primes given the list of gaps between them.

Compare the restored list with the original to ensure you did it right.

In [6]:
from sympy import primerange

ten_millionth_prime = 179424673

first_ten_million_primes = list(primerange(1, ten_millionth_prime + 1))

print(first_ten_million_primes[0:200])

def list_the_gaps(n=first_ten_million_primes):
    return list(map(lambda i: n[i]-n[i-1],range(2,len(n))))

def restore_primes(list_of_gaps):
    prime = 3
    plist = [2, 3]
    for gap in list_of_gaps:
        prime += gap
        plist.append(prime)
    return plist

list_of_gaps = list_the_gaps()
length_of_list_of_gaps = len(list_of_gaps)
print("Length of list of gaps:")
print(length_of_list_of_gaps)

KeyboardInterrupt: ignored

In [ ]:
first_ten_million_primes_restored = restore_primes(list_of_gaps)
first_ten_million_primes_restored == first_ten_million_primes